src/Compilers/Core/Portable/ReferenceManager/MergedAliases.cs

```diff
@@ -0,0 +1,100 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Collections.Immutable;
+using System.Diagnostics;
+
+namespace Microsoft.CodeAnalysis
+{
+    internal sealed class MergedAliases
+    {
+        public ArrayBuilder<string> AliasesOpt;
+        public ArrayBuilder<string> RecursiveAliasesOpt;
+
+        /// <summary>
+        /// Adds aliases of a specified reference to the merged set of aliases.
+        /// Consider the following special cases:
+        /// 
+        /// o {} + {} = {} 
+        ///   If neither reference has any aliases then the result has no aliases.
+        /// 
+        /// o {A} + {} = {A, global}
+        ///   {} + {A} = {A, global}
+        ///   
+        ///   If one and only one of the references has aliases we add the global alias since the 
+        ///   referenced declarations should now be accessible both via existing aliases 
+        ///   as well as unqualified.
+        ///   
+        /// o {A, A} + {A, B, B} = {A, A, B, B}
+        ///   We preserve dups in each alias array, but avoid making more dups when merging.
```

Comment:
Not really. Just preserving current behavior. If we want to dedup we should do so in a constructor of MetadataReferenceProperties. I'd consider that a separate work item.